client/finality-grandpa/src/finality_proof.rs

```diff
@@ -180,379 +128,177 @@ impl<B, Block: BlockT> FinalityProofProvider<B, Block>
 	pub fn new_for_service(
 		backend: Arc<B>,
 		storage_and_proof_provider: Arc<dyn StorageAndProofProvider<Block, B>>,
+		shared_authority_set: Option<SharedAuthoritySet<Block::Hash, NumberFor<Block>>>,
 	) -> Arc<Self> {
-		Arc::new(Self::new(backend, storage_and_proof_provider))
+		Arc::new(Self::new(
+			backend,
+			storage_and_proof_provider,
+			shared_authority_set,
+		))
 	}
 }
 
 impl<B, Block> FinalityProofProvider<B, Block>
-	where
-		Block: BlockT,
-		NumberFor<Block>: BlockNumberOps,
-		B: Backend<Block> + Send + Sync + 'static,
+where
+	Block: BlockT,
+	NumberFor<Block>: BlockNumberOps,
+	B: Backend<Block> + Send + Sync + 'static,
 {
-	/// Prove finality for the range (begin; end] hash. Returns None if there are no finalized blocks
-	/// unknown in the range.
-	pub fn prove_finality(
-		&self,
-		begin: Block::Hash,
-		end: Block::Hash,
-		authorities_set_id: u64,
-	) -> Result<Option<Vec<u8>>, ClientError> {
+	/// Prove finality for the given block number by returning a Justification for the last block of
+	/// the authority set.
+	pub fn prove_finality(&self, block: NumberFor<Block>) -> Result<Option<Vec<u8>>, ClientError> {
```

Comment:
The only downside of the current implementation is that we can't prove finality of blocks in the current set.

Reply:
Agree, a bit unfortunate. Do you think it's worth adding additional code paths for this purpose?

Reply:
I've added a return message to suggest using the subscription API in this case for now